pkg/ingester/ingester.go

```diff
@@ -2881,6 +2882,57 @@ func (i *Ingester) flushHandler(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusNoContent)
 }
 
+// ModeHandler Change mode of ingester. It will also update set unregisterOnShutdown to true if READONLY mode
+func (i *Ingester) ModeHandler(w http.ResponseWriter, r *http.Request) {
+	err := r.ParseForm()
+	if err != nil {
+		level.Warn(logutil.WithContext(r.Context(), i.logger)).Log("msg", "failed to parse HTTP request in mode handler", "err", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	currentState := i.lifecycler.GetState()
+	reqMode := strings.ToUpper(r.Form.Get("mode"))
+	switch reqMode {
+	case "READONLY":
+		if currentState != ring.READONLY {
+			err = i.lifecycler.ChangeState(r.Context(), ring.READONLY)
+			if err != nil {
+				respMsg := fmt.Sprintf("failed to change state: %s", err)
+				level.Warn(logutil.WithContext(r.Context(), i.logger)).Log("msg", respMsg)
+				w.WriteHeader(http.StatusBadRequest)
+				// We ignore errors here, because we cannot do anything about them.
+				_, _ = w.Write([]byte(respMsg))
+				return
+			}
+			i.lifecycler.SetUnregisterOnShutdown(true)
```

Comment:
What happens if we have a deployment when the ingester is in "READONLY?

Reply:
Good call out.
We do set this back when restarting so we dont lose the data, but talking offline to Alan his issue is we will not use this for query requests on restarting.
I will remove this feature then and change the steps to call shutdown when scaling down the ingesters. That will make sure to remove it from the ring.